### ISSUE TITLE
Make bulk/testport -b URL not only support official repositories, but any repository

### ISSUE THREAD

**Issue by michael-o** [freebsd/poudriere#1025]
## Prerequisites
- [X] Have you checked for an existing issue describing your idea?

## What is your proposal?
Remove the terms "official only" from this option and if any of the symbolic names, `latest`, `quarterly`, `release_X` keep repo name `FreeBSD` and enforce SRV RR, signatures, but when a custom URL is provided, set `FreeBSD` to `no`, add a new repo, e.g., `custom` where priority is higher (10?) and hardcode `mirror_type`, `pubkey`, `signature_type` to none. A more sophisticated way is to accept a filepath where the admin can supply custom `.conf` which contains arbitrary repos copied to `LOCALBASE` into the jail. In this case it needs to be clarified how to provide public keys, if required.

## What is the existing behavior, if any?
When providing the custom, selfhosted repo with `-b <URL>` the fetch fails because by packages aren't signed which I don't need nor is this an official repo.

## What is the motivation / use case for the change?
I build a base set of ports with custom options because the default ones do not work for our environment. Those options aren't only for direct, but also for automatic ones. On top are several other use cases (sets) which use those base ports partially as dependency which I don't want to build over and over again. The option for `bulk`/`testport` `-b <URL>` expects only official repos which quite limits this nice feature. I want to use it with any selfhosted repo which will satisfy the fetch request.

## Did you consider any alternatives?
Yes, patching `common.sh`, but this is an ugly hack. More specificly, in https://github.com/freebsd/poudriere/blob/5f69efba7802c412193cecf2d00a34fd77bb56fd/src/share/poudriere/common.sh#L3805-L3809 I has to switch of signature check and DNS SRV RRs to use my custom repo.

## Is this really a ports feature request?
Yes, it is a poudriere issue.

## Additional context
This was found in https://github.com/freebsd/poudriere/issues/883.

Here is a realworld example:
port set `ldadw_base`:
```
poudriere bulk -j 123-release-amd64 -z ldadw_base -O ldadw-custom -f /usr/local/etc/poudriere.d/123-release-amd64-default-ldadw_base-ports
```
with the following in `/usr/local/etc/poudriere.d/`:
```
123-release-amd64-default-ldadw_base-make.conf
123-release-amd64-default-ldadw_base-options/
123-release-amd64-default-ldadw_base-ports
```
`123-release-amd64-default-ldadw_base-ports`:
```
devel/git
devel/subversion-lts
devel/tig
editors/vim
editors/vim-colorschemes-legacy
ftp/curl
lang/python3
lang/python37
devel/py-pip
misc/mc
net/openldap26-client
security/krb5
shells/bash
shells/bash-completion
sysutils/htop
sysutils/msktutil
sysutils/tree
textproc/colordiff
textproc/rubygem-asciidoctor
net-mgmt/zabbix5-agent
security/siemens-cacerts
security/nss-siemens-cacerts-java
dns/register-hostnames
sysutils/system-update-periodic
```

Available internally at `https://deblndw011x.ad001.siemens.net/FreeBSD/packages/${ABI}/ldadw-base`

Now for a specific purpose bastille jail I have the following:
```
poudriere bulk -j 123-release-amd64 -z ldadocgen -O ldadw-custom -f /usr/local/etc/poudriere.d/123-release-amd64-default-ldadocgen-ports -b 'https://deblndw011x.ad001.siemens.net/FreeBSD/packages/${ABI}/ldadw-base'
```
with the following in `/usr/local/etc/poudriere.d/`:
```
123-release-amd64-default-ldadocgen-make.conf
123-release-amd64-default-ldadocgen-options/
123-release-amd64-default-ldadocgen-ports
```
`123-release-amd64-default-ldadocgen-ports`:
```
devel/apache-ant
java/openjdk8
math/gnuplot-lite
misc/getopt
net/py-ldap
www/apache24
www/mod_auth_gssapi
www/mod_wsgi4
www/py-flask
www/py-requests
net/py-urllib3
www/tomcat85
www/tomcat-native
x11-toolkits/libXmu
x11-fonts/fontconfig
print/ghostscript9-agpl-base
```
`123-release-amd64-default-ldadocgen-options/` contains merged options from `ldadw_base` and `ldadocgen` to satisfy `-b`'s requirements. Many of the ports in `ldadocgen` will use a lot from `ldadw_base`, e.g., apache24, tomcat85, tomcat-native will use apr which in `base` has been pulled by `subversion` already along with many other build dependencies , e.g., ninja, cmake, GNU autotools, etc.

**Comment by michael-o** [comment]
@bdrewery As promised, here is the issue to improve `-b`.

**Comment by michael-o** [comment]
@bdrewery gave it some thoughts today since I don't want to live with a patch forever. Here is my idea how we could move this forward: As far as I understand only a single repo is supported since `packagesite` is used throughout, would be much more work to rewrite to multiple `-b` options or supplying an external file. Compromise: If a URL is passed, don't threat is as official, disable `FreeBSD` repo and add a repo `custom` using that supplied URL.

WDYT? I'd be willing to work on a PR if this is acceptable for you...

**Comment by yannk** [comment]
I found this issue on my way to open my own.

I came up with the same type of fix:

```
        FreeBSD: {
                url: ${packagesite};
                mirror_type: NONE;
        }
```

I use `-b https://myownbinaryrepo` 

Seems like an easy fix would be to detect that `-b` is set to an URL and disable `SRV` in that case

**Comment by michael-o** [comment]
> I found this issue on my way to open my own.
> 
> I came up with the same type of fix:
> 
> ```
>         FreeBSD: {
>                 url: ${packagesite};
>                 mirror_type: NONE;
>         }
> ```
> 
> I use `-b https://myownbinaryrepo`
> 
> Seems like an easy fix would be to detect that `-b` is set to an URL and disable `SRV` in that case

That was my observation as well, but does not solve my wish situation where still only one repo is supported.

**Comment by dch** [comment]
what about add a var to `poudriere.conf` for this?

```
PACKAGE_REPO_FILE=/path/to/your/repo.conf
```

**Comment by michael-o** [comment]
> what about add a var to `poudriere.conf` for this?
> 
> ```
> PACKAGE_REPO_FILE=/path/to/your/repo.conf
> ```

Yes, this is one of the ways to solve the multirepo problem: https://github.com/freebsd/poudriere/issues/1025#issuecomment-1822979225. There is, though one problem. As written earlier, the entire setup is not prepared for that especially when bootstrapping pkg and sorting by priority...

**Comment by patmaddox** [comment]
#1239 provides a simple solution that works for me

**Comment by bdrewery** [comment]
I am going to merge #1271. I think it gets closer to resolving this issue? What is left?

**Comment by michael-o** [comment]
> I am going to merge [#1271](https://github.com/freebsd/poudriere/pull/1271). I think it gets closer to resolving this issue? What is left?

I think so, yes, but I will create a new issue for my larger usecase to track it: https://github.com/freebsd/poudriere/issues/1025#issuecomment-1822979225

**Comment by asomers** [comment]
I think there are really three things going on here:
1) Need to disable signature checking for private package servers
2) Need to the rquery bug when there are multiple package servers
3) Need to allow custom branch names

The first two are easy to solve.  I solve 1 just be writing a custom /etc/pkg/MYREPO.conf file.  It doesn't require any changes to Poudriere.  The file looks like this:
```
prod: {
  url: "http://MY-SERVER/pkg/${ABI}/prod",
  mirror_type: "none",
  signature_type: "none",
  enabled: yes
}
```

The second problem is also easy to solve, at least if you only want to fetch packages from one server.  In poudriere.conf, set `PACKAGE_FETCH_URL` accordingly, then apply the patch in #1323 .

The third problem requires a change to Poudriere.  It's not just a code change, more of a policy change.  IMHO I think Poudriere should just remove the branch name validation.
